File: PYTHON/STACK/generate_parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        result = []

        def backtrack(current: str, open_brackets: int, closed_brackets: int):
            if len(current) == 2 * n:
                result.append(current)
                return
            
            if open_brackets < n:
                backtrack(current + '(', open_brackets + 1, closed_brackets)
            
            if closed_brackets < open_brackets:
                backtrack(current + ')', open_brackets, closed_brackets + 1)

        backtrack('', 0, 0)

        return result

File: JAVASCRIPT/BACKTRACKING/generate_parentheses.js
function generateParenthesis(n) {
    const result = [];
    function backtrack(open, close, currentStr) {
        if (open === 0 && close === 0) {
            result.push(currentStr);
            return;
        }
        if (open > 0) backtrack(open-1, close, currentStr + '(');
        if (close > open) backtrack(open, close-1, currentStr + ')');
    }
    backtrack(n, n,  "");
    return result;
}